aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,196 @@
+
+Below is an annotated ``Movie`` model from an app that stores information about movies and TV shows. Most of the properties are required, the ``Id`` property is a primary key, and several string properties have length requirements. Additionally, there is a numeric range in place for the ``Price`` property from 0 to $999.99, along with a custom annotation.
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-34
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[CreditCard]``: Formats the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[Email]``: Formats the property with an email format.
+- ``[Phone]``: Formats the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Formats the property as a URL.
+
+A complete list of annotations is at the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_ namespace docs. Note that this namespace is not specific to MVC, but in a shared .NET namespace. 
+
+There may be intances where you need more features than built-in attributes provide. For those times, you can create custom validation attributes to apply to model properties.  
+
+Model State
+-----------
+`Model state <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Controller/index.html#prop-Microsoft.AspNet.Mvc.Controller.ModelState>`_ is a record of what data was bound to your model and the result of running model validation. If you have some logic that can produce a validation error, add that error to the model state's collection of errors by calling ``AddModelStateError`` as demonstrated below. Doing so will allow the the view to display the error.
+
+.. literalinclude:: validation/sample/MoviesController.cs
+   :language: c#
+   :lines: 18-39   
+   :emphasize-lines: 26, 27
+   :dedent: 8
+   
+After model binding and validation are complete, you may want to repeat parts of it. For example, a user may have entered text in a field expecting an integer. You can invoke validation again with the ``TryValidateModel`` method. 
```

Comment:
Suggest mentioning an example change to the model that an action might make before calling `TryValidateModel` e.g. parse the `string` for known values and set the `int` as appropriate.